lib/Sema/TypeChecker.h

```diff
@@ -491,6 +491,9 @@ enum TypeResolutionFlags : unsigned {
 
   /// Whether we are checking the underlying type of a typealias.
   TR_TypeAliasUnderlyingType = 0x4000000,
+
+  /// Whether we are checking the parameter list of a subscript.
+  TR_SubscriptParameters = 0x8000000,
```

Comment:
You could use this to resolve property types also, and remove the non-materializable check there.